File: python-flask/src/domain/order.py
from dataclasses import dataclass, field
from datetime import datetime, timezone
from uuid import uuid4


@dataclass
class OrderItem:
    product_id: str
    quantity: int
    price: float

    def subtotal(self) -> float:
        return self.quantity * self.price


@dataclass
class Order:
    order_id: str
    customer_id: str
    items: list[OrderItem]
    total_amount: float
    items_count: int
    created_at: datetime

    @staticmethod
    def create(customer_id: str, items: list[OrderItem]) -> "Order":
        total_amount = sum(item.subtotal() for item in items)
        items_count = sum(item.quantity for item in items)
        return Order(
            order_id=str(uuid4()),
            customer_id=customer_id,
            items=items,
            total_amount=total_amount,
            items_count=items_count,
            created_at=datetime.now(timezone.utc),
        )

File: go-stdlib/internal/domain/order.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type OrderItem struct {
	ProductID string
	Quantity  int
	Price     float64
}

func (oi OrderItem) Subtotal() float64 {
	return float64(oi.Quantity) * oi.Price
}

type Order struct {
	OrderID     string
	CustomerID  string
	Items       []OrderItem
	TotalAmount float64
	ItemsCount  int
	CreatedAt   time.Time
}

func NewOrder(customerID string, items []OrderItem) Order {
	var total float64
	var count int
	for _, item := range items {
		total += item.Subtotal()
		count += item.Quantity
	}
	return Order{
		OrderID:     uuid.New().String(),
		CustomerID:  customerID,
		Items:       items,
		TotalAmount: total,
		ItemsCount:  count,
		CreatedAt:   time.Now().UTC(),
	}
}
